src/backend/console.js

```diff
@@ -0,0 +1,85 @@
+// @flow
+
+import { getInternalReactConstants } from './renderer';
+import describeComponentFrame from './describeComponentFrame';
+
+import type { Fiber, ReactRenderer } from './types';
+
+let isDisabled: boolean = false;
+let hasPatched: boolean = false;
+
+export function disable(): void {
+  isDisabled = true;
+}
+
+export function enable(): void {
+  isDisabled = false;
+}
+
+const FRAME_REGEX = /\n {4}in /;
+
+export function patch(
+  targetConsole: Object,
+  { getCurrentFiber, findFiberByHostInstance, version }: ReactRenderer
+): void {
+  if (hasPatched) {
+    return;
+  }
+
+  if (typeof findFiberByHostInstance !== 'function') {
+    return;
+  }
+
+  const { getDisplayNameForFiber } = getInternalReactConstants(version);
+
+  hasPatched = true;
+
+  for (let method in targetConsole) {
+    const appendComponentStack =
+      typeof getCurrentFiber === 'function' &&
+      (method === 'error' || method === 'warn' || method === 'trace');
+
+    const originalMethod = targetConsole[method];
+    const overrideMethod = (...args) => {
+      if (isDisabled) return;
+
+      if (appendComponentStack) {
+        // If we are ever called with a string that already has a component stack, e.g. a React error/warning,
+        // don't append a second stack.
+        const alreadyHasComponentStack =
+          args.length > 0 && FRAME_REGEX.exec(args[args.length - 1]);
+
+        if (!alreadyHasComponentStack) {
+          // $FlowFixMe We know getCurrentFiber() is a function if appendComponentStack is true.
+          let current: ?Fiber = getCurrentFiber();
+          let ownerStack: string = '';
+          while (current != null) {
+            const name = getDisplayNameForFiber(current);
+            const owner = current._debugOwner;
+            const ownerName =
+              owner != null ? getDisplayNameForFiber(owner) : null;
+
+            ownerStack += describeComponentFrame(
+              name,
+              current._debugSource,
+              ownerName
+            );
+
+            current = owner;
+          }
+
+          if (ownerStack !== '') {
+            args.push(ownerStack);
+          }
+        }
+      }
+
+      originalMethod(...args);
+    };
+
+    try {
+      // $FlowFixMe property error|warn is not writable.
+      targetConsole[method] = overrideMethod;
```

Comment:
One thing to be aware of is that this might cause the methods to lose location information displayed by the browser. Like when you warn/error and it says "foo.js:XX" in the upper right corner:

<img width="490" alt="Screen Shot 2019-07-16 at 23 02 10" src="https://user-images.githubusercontent.com/810438/61333076-c2f70f80-a81d-11e9-84a9-b6d5cebf3e5b.png">

I guess it's kind of unavoidable. Although would be nice to start a conversation with browsers if we can somehow specify to skip the topmost stack level when displaying it.

(We have a similar problem in CRA already because we also override it there.)

Reply:
Yeah, I agree with this as a downside. I will reach out to some folks on Chrome's developer tools team to at least start that conversation.

Reply:
One way to mitigate or limit impact from this would be to have DevTools call the Global Hook when it begins or exits the React work loop. Then you could only inject the overrides at that point, and restore them afterwards.

Reply:
Hm... do you mean the work loop would call a function on the DevTools global hook (or otherwise notify it) any time it started/stopped work?